### PR TITLE
gh-283: set up nox for developers and CI

### PR DESCRIPTION
**Pull request by Saransh-cpp**
- Adds `noxfile` to support `nox` sessions for testing, docs, and linting/formatting.
- `Nox` is now used to run everything in the CI.
- Adds documentation about the new infrastructure in contributing guidelines.

Closes: #283

### CODE REVIEW

**Comment by paddyroddy** [comment]
Blocked by #289

**Comment by paddyroddy** [comment]
CI now running following #289

**Comment by Saransh-cpp** [comment]
Ahh that was a pain, finally works and ready for a review!

File: coverage.xml [hunk 1]
```diff
@@ -0,0 +1,942 @@
+<?xml version="1.0" ?>
```
Comment (paddyroddy) [inline coverage.xml:1]: You presumably didn't mean to add this file?

File: noxfile.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+from __future__ import annotations
+
+import nox
+
+# Options to modify nox behaviour
+nox.options.default_venv_backend = "uv|virtualenv"
+nox.options.reuse_existing_virtualenvs = True
+nox.options.sessions = ["lint", "tests"]
+
+ALL_PYTHON = ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
Comment (paddyroddy) [inline noxfile.py:12]: Seems silly but then only one line change when we drop support for one
```suggestion
ALL_PYTHON = [
	"3.8",
	"3.9",
	"3.10",
	"3.11",
	"3.12"
]
```

File: noxfile.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        "pytest",
+        *session.posargs,
+    )
+
+
+@nox.session(python=ALL_PYTHON)
+def coverage(session: nox.Session) -> None:
+    """Run tests and compute coverage."""
+    session.posargs.append("--cov=glass")
+    tests(session)
```
Comment (paddyroddy) [inline noxfile.py:36]: Is `tests` nox magic?

File: noxfile.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    session.install("-e", ".[examples]")
+    session.run("jupyter", "execute", "examples/**/*.ipynb", *session.posargs)
+
+
+@nox.session
+def docs(session: nox.Session) -> None:
+    """Build the docs. Pass "serve" to serve."""
+    session.install("-e", ".[docs]")
+    session.chdir("docs")
+    session.run("sphinx-build", "-M", "html", ".", "_build")
```
Comment (paddyroddy) [inline noxfile.py:59]: I hate `sphinx` so much lol

File: noxfile.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    if session.posargs:
+        if "serve" in session.posargs:
+            print("Launching docs at http://localhost:8001/ - use Ctrl-C to quit")
+            session.run("python", "-m", "http.server", "8001", "-d", "_build/html")
```
Comment (paddyroddy) [inline noxfile.py:64]: Can you create a variable for the port and re-use it?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -138,6 +138,52 @@ long as there have been changes.
...
+```
+
+Only `tests`, `coverage`, and the `doctests` session run on all supported Python
+versions by default.
+
+To specify a particular Python version (for example `3.11`), use the following
+syntax -
+
+```bash
+nox -s tests-3.11
```
Comment (paddyroddy) [inline CONTRIBUTING.md:175]: Presumably, it is trivial to test on different OS with `nox` too?

**Review by paddyroddy — CHANGES_REQUESTED** [review]
This looks great, nice documentation too. Some comments and questions.

File: pyproject.toml [hunk 7]
```diff
@@ -55,6 +55,7 @@ examples = [
     "matplotlib",
 ]
 test = [
+    "coverage",
```
Comment (Saransh-cpp) [inline pyproject.toml:58]: This should not be here
```suggestion
```

File: coverage.xml (hunk 1 above)
Comment (Saransh-cpp) [reply coverage.xml:1]: Yes, thanks!

File: noxfile.py (hunk 3 above)
Comment (Saransh-cpp) [reply noxfile.py:36]: It is the nox session defined above. The `coverage` session just passes an extra argument to the `tests` session.

File: CONTRIBUTING.md (hunk 6 above)
Comment (Saransh-cpp) [reply CONTRIBUTING.md:175]: That should be just a matter of adding `if-else` within the sessions. For example, the tests session could be -
```py
@nox.session(python=ALL_PYTHON)
 def tests(session: nox.Session) -> None:
    """Run the unit tests."""
    session.install("-c", ".github/test-constraints.txt", "-e", ".[test]")
    if sys.platform != "win32":
        session.install("library not supported by windows")  # pytest should handle rest (`importorskip`, `skipif`, ...)
    session.run(
       "pytest",
       *session.posargs,
    )

File: noxfile.py (hunk 4 above)
Comment (Saransh-cpp) [reply noxfile.py:59]: Bahaha, is there an alternative that I can use here?

File: noxfile.py (hunk 3 above)
Comment (paddyroddy) [reply noxfile.py:36]: Ah I see, thanks

File: noxfile.py (hunk 4 above)
Comment (paddyroddy) [reply noxfile.py:59]: Leave it as it is. I just tend to use https://github.com/machow/quartodoc, https://github.com/mkdocs/mkdocs, etc.

File: CONTRIBUTING.md (hunk 6 above)
Comment (paddyroddy) [reply CONTRIBUTING.md:175]: Okay. Am intrigued why you prefer `nox` over `tox` 😉 but we can talk about this offline

**Review by paddyroddy — APPROVED** [review]
🥇

**Pull request merged**